File: 2024/Day_01/Go/day_01.go
// Solution to Advent of Code 2024 Day 01 - Historian Hysteria

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

// MultipleLines returns the input file if it contains multiple lines of text
func MultipleLines(fileName string) ([]string, error) {
	inputSlice := make([]string, 0)
	file, err := os.Open(fileName)
	if err != nil {
		return inputSlice, err
	}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		inputSlice = append(inputSlice, scanner.Text())
	}
	err = file.Close()
	if err != nil {
		return inputSlice, err
	}
	return inputSlice, nil
}

func main() {
	lists, err := MultipleLines("../input.txt")
	if err != nil {
		fmt.Printf("Error%v\n", err)
	}
	var leftSide []float64
	var rightSide []float64
	// split the left and right and convert them to ints.
	for _, line := range lists {
		temp := strings.Fields(line)
		left, _ := strconv.ParseFloat(temp[0], 64)
		right, _ := strconv.ParseFloat(temp[1], 64)
		leftSide = append(leftSide, left)
		rightSide = append(rightSide, right)
	}
	// Take advantage of the new slice sorting in Go. It sorts in place
	slices.Sort(leftSide)
	slices.Sort(rightSide)

	// a new slice to hold the differences
	var diff []float64

	for index, value := range leftSide {
		diff = append(diff, math.Abs(value-rightSide[index]))
	}

	var sum float64

	for _, value := range diff {
		sum += value
	}

	fmt.Printf("The total distance between the lists is %f\n", sum)

	// map to hold a count of values in the right side
	var counter = make(map[float64]int)

	for _, number := range rightSide {
		counter[number] += 1
	}

	var product []float64

	for _, number := range leftSide {
		product = append(product, number*float64(counter[number]))
	}

	var productSum float64

	for _, number := range product {
		productSum += number
	}

	fmt.Printf("The simlarity score is %f", productSum)
}

File: 2024/Day_01/Python/day_01.py
from collections import Counter
def input_per_line(file: str):
    """This is for when each line is an input to the puzzle. The newline character is stripped."""
    with open(file, 'r') as input_file:
        return [line.rstrip() for line in input_file.readlines()]

if __name__ == '__main__':
    locations = input_per_line("../input.txt")
    left_side = []
    right_side = []
    for line in locations:
        entries = line.split()
        left_side.append(int(entries[0]))
        right_side.append(int(entries[1]))
    left_side.sort()
    right_side.sort()
    distances = [abs(right_side[number]-left_item) for number, left_item in enumerate(left_side)]
    print(f"The total distance between the lists is {sum(distances)}")
    right_side_count = Counter(right_side)
    similarity_list = [item * right_side_count[item] for item in left_side]
    print(f"The similarity score is {sum(similarity_list)}")
